src/ssl.c

```diff
@@ -54563,65 +54563,96 @@ int wolfSSL_BN_rand(WOLFSSL_BIGNUM* bn, int bits, int top, int bottom)
     WC_RNG        tmpRNG[1];
     byte          buff[1024];
 #endif
+    int           lastBits = 0;
 
-    (void)top;
-    (void)bottom;
-    WOLFSSL_MSG("wolfSSL_BN_rand");
+    WOLFSSL_ENTER("wolfSSL_BN_rand");
 
-    if (bits <= 0) {
-        return WOLFSSL_FAILURE;
+    if ((bn == NULL || bn->internal == NULL) || bits < 0 ||
+        (bits == 0 && (bottom != 0 || top != -1)) || (bits == 1 && top > 0)) {
+        WOLFSSL_MSG("Bad argument");
+        ret = WOLFSSL_FAILURE;
     }
 
-    len = bits / 8;
-    if (bits % 8)
-        len++;
+    if (ret == WOLFSSL_SUCCESS) {
+        lastBits = bits % 8;
+        if (lastBits) {
+            len++;
+        }
 
-    /* has to be a length of at least 1 since we set buf[0] and buf[len-1] */
-    if (len < 1) {
-        return WOLFSSL_FAILURE;
+    #ifdef WOLFSSL_SMALL_STACK
+        buff   = (byte*)XMALLOC(1024,        NULL, DYNAMIC_TYPE_TMP_BUFFER);
+        tmpRNG = (WC_RNG*) XMALLOC(sizeof(WC_RNG), NULL, DYNAMIC_TYPE_RNG);
+        if (buff == NULL || tmpRNG == NULL) {
+            XFREE(buff,   NULL, DYNAMIC_TYPE_TMP_BUFFER);
+            XFREE(tmpRNG, NULL, DYNAMIC_TYPE_RNG);
+            ret = WOLFSSL_FAILURE;
+        }
+    #endif
     }
 
-#ifdef WOLFSSL_SMALL_STACK
-    buff   = (byte*)XMALLOC(1024,        NULL, DYNAMIC_TYPE_TMP_BUFFER);
-    tmpRNG = (WC_RNG*) XMALLOC(sizeof(WC_RNG), NULL, DYNAMIC_TYPE_RNG);
-    if (buff == NULL || tmpRNG == NULL) {
-        XFREE(buff,   NULL, DYNAMIC_TYPE_TMP_BUFFER);
-        XFREE(tmpRNG, NULL, DYNAMIC_TYPE_RNG);
-        return ret;
+    if (ret == WOLFSSL_SUCCESS) {
+        if (wc_InitRng(tmpRNG) == 0) {
```

Comment:
Not sure why we are using a new RNG.
Creating an RNG is time consuming and uses entropy.
OpenSSL uses the global I believe.